# คลาสและอ็อบเจกต์

Java เป็นภาษาที่ออกแบบมาโดยมุ่งเน้นที่กระบวนทัศน์การโปรแกรมเชิงวัตถุ (Object-Oriented Programming [OOP]) ซึ่งเป็นรูปแบบการพัฒนาโปรแกรมที่มองโปรแกรมเป็นการประกอบกันของหน่วยย่อยที่เรียกว่าอ็อบเจกต์ (object) ซึ่งประกอบด้วยข้อมูลหรือคุณลักษณะ (attribute) และรหัสคำสั่งในรูปของเมทอด (method) โดยอ็อบเจกต์เหล่านี้จะมีหน้าที่แตกต่างกันไปและทำงานร่วมกัน

อ็อบเจกต์หนึ่ง ๆ จะมีคุณลักษณะและพฤติกรรมในการทำงานตามรูปแบบที่ถูกกำหนดโดยคลาส (class) ที่เป็นต้นแบบของอ็อบเจกต์ อ็อบเจกต์จะถูกสร้างขึ้นจากคลาส ซึ่งหมายความว่าการจะสร้างอ็อบเจกต์ขึ้นมาได้ ต้องมีการกำหนดคลาสขึ้นมาก่อน คลาสจะเป็นตัวกำหนดว่าอ็อบเจกต์ที่จะถูกสร้างขึ้นนั้นจะประกอบไปด้วยคุณลักษณะใดบ้าง โดยกำหนดในรูปของตัวแปรและข้อมูลต่าง ๆ ที่อ็อบเจกต์จะมี และสามารถทำอะไรได้บ้าง โดยกำหนดในรูปของเมทอด

ตัวอย่างต่อไปนี้เป็นโค้ดของคลาสชื่อว่า `Pair` ซึ่งกำหนดให้อ็อบเจกต์ที่จะสร้างขึ้นจากคลาสนี้มีข้อมูลหรือคุณลักษณะในรูปของตัวแปร `first` และ `second` และมีเมทอด `getFirst`, `getSecond`, `setFirst`, `swap` และ `print` ให้เรียกใช้งานได้

In [3]:
public class Pair {
    private int first;
    private int second;
    
    public int getFirst() {
        return first;
    }
    
    public int getSecond() {
        return second;
    }
    
    public void setFirst(int value) {
        first = value;
    }
    
    public void swap() {
        int temp = first;
        first = second;
        second = temp;
    }
    
    public void print() {
        System.out.println("(" + first + ", " + second + ")");
    }
}

## ความสัมพันธ์ระหว่างคลาสกับอ็อบเจกต์

การทำความเข้าใจถึงความสัมพันธ์ระหว่างคลาสกับอ็อบเจกต์เป็นเรื่องสำคัญสำหรับผู้ที่เริ่มต้นศึกษาแนวคิด OOP เพราะทั้ง 2 สิ่งนี้คือองค์ประกอบหลักของโปรแกรมแบบ OOP และมีรายละเอียดบางอย่างที่เฉพาะเจาะจงสำหรับคลาสหรืออ็อบเจกต์ที่อาจสร้างความสับสนได้ง่ายหากความเข้าใจในประเด็นนี้ยังไม่ชัดเจน

เราอาจจะมองความสัมพันธ์ระหว่างคลาสกับอ็อบเจกต์ได้ใน 2 ลักษณะดังนี้

### คลาสเป็นพิมพ์เขียวของอ็อบเจกต์

คลาสเป็นพิมพ์เขียว (blueprint) หรือแบบ (design) ที่กำหนดลักษณะต่าง ๆ ของอ็อบเจกต์ ในมุมมองนี้ เราเปรียบเทียบคลาสและอ็อบเจกต์กับกระบวนการผลิตอื่น ๆ ในโลกความจริง เช่น รถยนต์ ในการผลิตรถยนต์ เราจะเริ่มต้นจากขั้นตอนการออกแบบก่อน ซึ่งในแบบนั้นเราจะกำหนดลักษณะต่าง ๆ ว่ารถยนต์ที่จะผลิตนี้จะประกอบด้วยอะไรบ้าง มีกี่ล้อ กี่ประตู เครื่องยนต์กี่ลูกสูบ ขนาดเท่าใด ระบบเกียร์แบบไหน มีฟังก์ชันใดบ้าง (นอกจากการขับเคลื่อนปกติ) แต่ละฟังก์ชันนั้นทำงานได้อย่างไร ลึกไปจนถึงระดับของกลไกและวงจรที่ควบคุมการทำงานเหล่านี้ จนได้ออกมาเป็นแบบหรือพิมพ์เขียว

เมื่อเราได้แบบแล้ว เราจึงสามารถผลิตรถยนต์จริง ๆ ออกมาได้ โดยรถยนต์แต่ละคันก็จะมีลักษณะตามแบบเหมือน ๆ กัน แบบหนึ่งแบบสามารถนำไปสร้างรถยนต์กี่คันก็ได้ตามความต้องการ และถึงแม้รถยนต์ที่สร้างขึ้นแต่ละคันจะมีลักษณะตามแบบเหมือน ๆ กัน ก็ยังมีลักษณะเฉพาะบางอย่างที่แตกต่างกันไปในแต่ละคันได้ เช่น หมายเลขตัวถัง สี ออปชันเสริม เป็นต้น ซึ่งในแบบเองก็จะระบุไว้เช่นกันว่าลักษณะเฉพาะใดบ้างที่สามารถกำหนดให้แตกต่างได้ และในการใช้งานจริง รถแต่ละคันก็จะทำงานแยกกันไปโดยอิสระ

ในมุมมองนี้ เราเปรียบเทียบคลาสเป็นแบบ และอ็อบเจกต์เป็นรถยนต์แต่ละคัน ตัวแบบไม่สามารถนำมาใช้งานได้ เราไม่สามารถขับแบบของรถยนต์ได้ เราต้องขับรถยนต์จริง ๆ เช่นเดียวกัน เราไม่สามารถใช้งานคลาสตรง ๆ ได้ (ยกเว้นในบางกรณีที่เราจะพูดถึงต่อไป) แต่เราต้องสร้างอ็อบเจกต์ขึ้นมาเพื่อใช้งาน โดยเราสร้างอ็อบเจกต์จากคลาส คลาสหนึ่งคลาสสามารถนำไปสร้างอ็อบเจกต์กี่อ็อบเจกต์ก็ได้แล้วแต่ความต้องการใช้งาน และอ็อบเจกต์แต่ละตัวที่เราสร้างขึ้นจากคลาสเดียวกันก็จะมีลักษณะและพฤติกรรมแบบเดียวกัน แต่ทำงานอิสระต่อกัน

รถยนต์คันหนึ่ง ๆ ที่ผลิตออกมาจะมีคุณลักษณะบางอย่างที่ถูกกำหนดขึ้นตั้งแต่ขั้นตอนการออกแบบและผลิต จำนวนล้อ เครื่องยนต์ที่ใช้ ระบบเกียร์ ถูกกำหนดตั้งแต่ขั้นตอนการออกแบบ และสี หมายเลขตัวถัง ถูกกำหนดในขั้นตอนการผลิต คุณลักษณะเหล่านี้เป็นคุณลักษณะที่อยู่ติดตัวและไม่เปลี่ยนแปลง แต่เมื่อใดที่รถยนต์ถูกนำไปใช้งาน จะมีคุณลักษณะเฉพาะอีกชุดหนึ่งของรถยนต์แต่ละคันที่จะแปรผันไปตามการใช้งาน เช่น ความเร็ว รอบเครื่องยนต์ เกียร์ที่กำลังใช้ เป็นต้น คุณลักษณะเหล่านี้มีลักษณะเป็นพลวัต มีค่าไม่คงที่ แปรเปลี่ยนได้ทุกขณะตามสภาวะการทำงานของรถยนต์ คุณลักษณะเหล่านี้ถูกพิจารณาตั้งแต่ในขั้นตอนการออกแบบในรูปของตัวแปรซึ่งมีผลต่อการคำนวณในกระบวนการทำงานของฟังก์ชันต่าง ๆ ของรถยนต์

เมื่อเราพิจารณาในรูปของโค้ดซึ่งประกอบด้วยคลาสและอ็อบเจกต์ดังนี้

In [4]:
public class MiniTruck {
    private final Wheel[] wheels = new Wheels[4];
    private final Engine engine = new Engine("V4 Diesel Engine");
    private final ManualGearbox gearbox = new ManualGearbox();

    private final String chassisNumber;
    private final String color;
    
    private int gearPosition;
    private double speed;
    private double rpm;
    
    public MiniTruck(String chassisNumber, String color) {
        this.chassisNumber = chassisNumber;
        this.color = color;
        
        gearPosition = 1;
        speed = 0.0;
        rpm = 0.0;
    }
    
    public void changeGear(int gearPosition) {
        this.gearPosition = gearPosition;
    }
    
    public void accelerate(double rpmChange) {
        rpm += rpmChange;
        speed = gearPosition * rpm / 100.0;
    }
    
    // ... code omitted
    
    public static void main(String[] args) {
        MiniTruck blackTruck = new MiniTruck("JM6-123456-11111111", "black");
        MiniTruck whiteTruck = new MiniTruck("JM6-123456-22222222", "white");

        blackTruck.accelerate(500.0);
        whiteTruck.changeGear(2);
        whiteTruck.accelerate(800.0);
    }
}

โค้ดในตัวอย่างนี้อาจจะมีหลายส่วนที่เรายังไม่ได้อธิบาย แต่ส่วนเหล่านั้นยังไม่ใช่ประเด็นสำคัญในที่นี้ ในโค้ดนี้เราจะเห็นคุณลักษณะของ `MiniTruck` ในรูปของตัวแปรต่าง ๆ ได้แก่ `wheels`, `engine`, `gearbox`, `chassisNumber`, `color`, `gearPosition`, `speed` และ `rpm` ตัวแปรทั้งหมดนี้ประกาศขึ้นโดยคลาส `MiniTruck` เพื่อบอกว่าเมื่อเราสร้างอ็อบเจกต์จากคลาสนี้ขึ้นมา อ็อบเจกต์นั้นจะมีคุณลักษณะอะไรบ้าง

ล้อ (`wheels`) เครื่องยนต์ (`engine`) และระบบเกียร์ (`gearbox`) ถูกกำหนดไว้ตายตัวในโค้ดของคลาสเลย เปรียบได้กับการกำหนดไว้ตั้งแต่ตอนออกแบบ หมายเลขตัวถัง (`chassisNumber`) และสีรถ (`color`) จะถูกกำหนดตอนที่สร้างอ็อบเจกต์โดยส่งมาเป็นอาร์กิวเมนต์ตอนสร้าง เทียบได้กับการกำหนดในขั้นตอนการผลิต และตำแหน่งเกียร์ (`gearPosition`) ความเร็ว (`speed`) และรอบเครื่องยนต์ (`rpm`) จะเป็นค่าที่ใช้งานตอนที่อ็อบเจกต์ถูกสร้างขึ้นมาแล้ว แต่ละอ็อบเจกต์ที่ถูกสร้างขึ้นมา ซึ่งในที่นี้คือ `blackTruck` และ `whiteTruck` ก็จะมีค่าเหล่านี้เป็นของตัวเองแยกกันไป

คลาสจะเป็นตัวกำหนดคุณลักษณะและการทำงานของอ็อบเจกต์แต่ละตัว อ็อบเจกต์แต่ละตัวก็จะมีคุณลักษณะเหล่านี้ของตัวเองแยกกันไปไม่ขึ้นต่อกัน

กลับไปที่ตัวอย่าง `Pair` อีกครั้ง เมื่อเราลองสร้างและเรียกใช้งานอ็อบเจกต์จากคลาส `Pair`

In [2]:
Pair p1 = new Pair();
Pair p2 = new Pair();

p1.setFirst(1);

p2.setFirst(3);
p2.swap();
p2.setFirst(2);

p1.print();
p2.print();

(1, 0)
(2, 3)


จากตัวอย่างด้านบนนี้ เราจะเห็นว่าอ็อบเจกต์ที่ถูกสร้างขึ้น (ด้วยคำสั่ง `new`) มี 2 ตัว คือ `p1` และ `p2` และถึงแม้ทั้ง 2 อ็อบเจกต์นี้จะมีตัวแปร `first` และ `second` เป็นคุณลักษณะเหมือนกัน แต่ก็เป็นของใครของมัน ไม่ได้มีค่าเดียวกัน โดยเมื่อสิ้นสุดการทำงาน `p1` มีค่าของ `first` และ `second` เป็น 1 และ 0 ตามลำดับ และ `p2` มีค่าเป็น 2 และ 3 ตามลำดับ

### คลาสเป็นกลุ่มชนิดของอ็อบเจกต์

"มนุษย์" เป็นมโนทัศน์ (concept) ที่ใช้เรียกของกลุ่มชนิดของสิ่งมีชีวิตประเภทหนึ่ง โดยตัวคำว่ามนุษย์เองไม่ได้หมายถึงบุคคลใดบุคคลหนึ่ง แต่เป็นตัวแทนของบุคคลบนโลกทั้งหมด เช่นเดียวกัน คำว่า "สุนัข" ก็ไม่ได้หมายถึงสุนัขตัวใดตัวหนึ่ง แต่เป็นคำเรียกแทนสุนัขทั้งหมด เมื่อเราต้องการระบุถึงมนุษย์คนใดคนหนึ่งหรือสุนัขตัวใดตัวหนึ่ง เราจะมีคำอ้างอิงในลักษณะของชื่อ เช่น วรุตม์ แจ็ก กอลลัม สรรพนาม เช่น เขา มัน หรือคำบ่งชี้ เช่น คนนั้น คนนี้ ตัวนั้น เป็นต้น

ในมุมมองนี้ เราจะเห็นว่ากลุ่มชนิดเป็นมโนทัศน์หรือแนวคิดเชิงนามธรรม เป็นการตั้งชื่อเรียกกลุ่มของสิ่งที่เป็นรูปธรรมที่มีลักษณะบางอย่างร่วมกัน มนุษย์ทุกคนมีคุณลักษณะเหมือน ๆ กัน มีหน้าตา สีผิว เพศ มีแขน ขา มีนิ้ว มีสติปัญญา และมีคุณลักษณะและความสามารถอื่น ๆ ที่มนุษย์ต่างก็ทำได้เหมือนกัน ซึ่งสิ่งเหล่านี้กำหนดความเป็นมนุษย์ แต่ความเฉพาะของคุณลักษณะเหล่านั้นทำให้มนุษย์แต่ละคนแตกต่างกันออกไป ทุกคนมีลักษณะของหน้าตาต่างกัน มีสีผิวต่างกัน มีเพศต่างกัน เป็นต้น เทียบกับคลาสและอ็อบเจกต์ คำว่ามนุษย์ก็เหมือนคลาสหนึ่ง ซึ่งกำหนดว่าการเป็นมนุษย์จะต้องมีคุณลักษณะอะไรบ้าง แต่มนุษย์แต่ละคน หรืออ็อบเจกต์แต่ละอ็อบเจกต์ จะมีค่าเฉพาะของคุณลักษณะเหล่านั้นที่แตกต่างกันไป

และเมื่อเราพูดถึง "รถ" เราหมายถึงยานพาหนะทางบกที่สามารถบรรทุกคนหรือสิ่งของจากจุดหนึ่งไปสู่อีกจุดหนึ่งได้ ความสามารถในการบรรทุกคนหรือสิ่งของ การเป็นยานพาหนะทางบก และการเคลื่อนที่ได้ เป็นความสามารถของรถทุกคัน แต่รถแต่ละคันอาจจะมีลักษณะเฉพาะของตัวเองได้ เช่น สี จำนวนล้อ จำนวนประตู และเมื่อขับเคลื่อนก็มีคุณลักษณะแบบพลวัตของตน เช่น ความเร็ว ทิศทาง เป็นต้น

การมองคลาสเป็นการจำแนกกลุ่มนี้ ทำให้เห็นว่าการกระทำที่เกิดขึ้นจริงใด ๆ จะเป็นการกระทำต่อตัวตนที่เป็นรูปธรรมแต่ละตัวในกลุ่มชนิดนั้น ๆ ซึ่งก็คืออ็อบเจกต์ เราไม่สามารถกระทำต่อคลาสได้เนื่องจากคลาสเป็นนามธรรม เราไม่สามารถเล่นกับ "สุนัข" ในความหมายของการจำแนกกลุ่มหรือที่เป็นมโนทัศน์ แต่เราสามารถเล่นกับสุนัข "ตัวนี้" ได้ หรือเล่นกับ "เจ้ากอลลัม" ได้ เราไม่สามารถพูดคุยกับ "มนุษย์" ในความหมายของกลุ่มชนิดของสิ่งมีชีวิตที่เรียกว่ามนุษย์ แต่เราสามารถพูดคุยกับมนุษย์ "คนนั้น" หรือ "วรุตม์" ได้ เราอาจจะพูดว่า "มนุษย์" สามารถขับ "รถ" ได้ แต่การพูดเช่นนั้นเป็นเพียงการบอกถึงความสามารถของมนุษย์ในภาพรวม เหตุการณ์การขับรถใด ๆ ที่เกิดขึ้นจริงจะต้องเกิดขึ้นจากมนุษย์ "คนนั้น" ขับรถ "คันนั้น"

คลาสในมุมมองนี้จะเป็นตัวบ่งบอกว่าสมาชิกของกลุ่มชนิดนี้ ซึ่งก็คืออ็อบเจกต์ที่สร้างจากคลาสนี้ จะมีคุณลักษณะอย่างไร สามารถทำอะไรได้บ้าง กระทำต่อคลาสอื่นใดบ้าง แต่การกระทำที่เกิดขึ้นจริง ๆ จะต้องเกิดขึ้นกับอ็อบเจกต์ ไม่ได้เกิดขึ้นกับคลาส

นอกจากนี้ คลาสในมุมมองของกลุ่มชนิดยังแฝงนัยของการมีกลุ่มชนิดย่อยด้วย เราสามารถแบ่งย่อย "มนุษย์" ลงไปเป็น "ผู้ชาย" และ "ผู้หญิง" เช่นเดียวกับที่เราอาจแบ่งย่อย "พนักงาน" เป็น "พนักงานประจำ" และ "พนักงานชั่วคราว" ได้ "พนักงานชั่วคราว" ก็อาจแบ่งย่อยลงไปเป็น "พนักงานจ้างรายวัน" และ "พนักงานจ้างรายชั่วโมง" ได้อีก การแบ่งย่อยนี้ กลุ่มชนิดย่อยจะมีคุณลักษณะและความสามารถของกลุ่มชนิดรวม แต่จะมีลักษณะเฉพาะที่แตกต่างออกไปในแต่ละกลุ่มชนิดย่อยด้วย "ผู้ชาย" และ "ผู้หญิง" ต่างก็มีคุณลักษณะและความสามารถของ "มนุษย์" เหมือนกัน แต่มีลักษณะเฉพาะอื่น ๆ ที่แตกต่างกัน "พนักงานประจำ" "พนักงานจ้างรายวัน" และ "พนักงานจ้างรายชั่วโมง" ต่างก็มีคุณลักษณะของ "พนักงาน" เหมือนกัน แต่ก็มีรายละเอียดอื่น ๆ ที่แตกต่างกัน และในกรณีของ "พนักงานจ้างรายวัน" และ "พนักงานจ้างรายชั่วโมง" นอกจากจะมีคุณลักษณะของ "พนักงาน" เหมือนกันแล้ว ยังมีคุณลักษณะของ "พนักงานชั่วคราว" ร่วมกันด้วย

และเมื่อเรากล่าวถึงสมาชิก (อ็อบเจกต์) อันใดอันหนึ่งของกลุ่มชนิดย่อย สมาชิกนั้นจะเป็นสมาชิกของกลุ่มชนิดรวมด้วย "วรุตม์" อยู่ในกลุ่มชนิด "ผู้ชาย" หมายความว่า "วรุตม์" อยู่ในกลุ่มชนิด "มนุษย์" ด้วยเช่นกัน หรือถ้า "วรุตม์" เป็น "พนักงานจ้างรายวัน" ก็หมายถึง "วรุตม์" เป็น "พนักงานชั่วคราว" และ "พนักงาน" ด้วย

ลักษณะนี้เป็นลักษณะที่สอดคล้องกับคุณสมบัติของการโปรแกรมเชิงวัตถุ ได้แก่ การสืบทอด (inheritance) และการมีหลายรูปแบบ (polymorphism) ซึ่งเราจะกล่าวถึงในภายหลัง

### อ็อบเจกต์เป็นตัวตนของคลาส

จากความสัมพันธ์ระหว่างคลาสกับอ็อบเจกต์ที่เราได้กล่าวถึงไปแล้วนั้น เราจะเห็นว่าคลาสไม่สามารถทำงานได้ด้วยตัวเอง (ยกเว้นในกรณีเฉพาะบางกรณี) การทำงานจริงนั้นจะเกิดขึ้นจากอ็อบเจกต์ซึ่งต้องสร้างขึ้นโดยมีคลาสเป็นแบบหรือพิมพ์เขียว เรากล่าวได้ว่าอ็อบเจกต์เป็นตัวตน (instance) ของคลาส

> **หมายเหตุ** พจนานุกรมศัพท์คอมพิวเตอร์และเทคโนโลยีสารสนเทศ ฉบับราชบัณฑิตยสถาน พ.ศ. 2546 กำหนดให้คำว่า instance แปลว่า กรณีตัวอย่าง แต่ในที่นี้เพื่อความหมายที่น่าจะชัดเจนขึ้น ขอใช้คำว่า ตัวตน (embodiment) แทนความหมายของคำว่า instance

### การประกาศคลาส

ในเบื้องต้น การประกาศคลาสมีรูปแบบวากยสัมพันธ์ดังนี้

```
<class-modifiers> class ClassName {
    <field-declarations>
    <constructor-declarations>
    <method-declarations>
}
```

ส่วนประกอบต่าง ๆ ในการประกาศคลาสมีความหมายดังนี้

- `<class-modifiers>` เป็นตัวกำหนดคุณสมบัติบางประการของคลาส เช่น `public`, `abstract` หรือ `final` โดยในเบื้องต้นเราจะใช้ตัวกำหนด `public` หรือละไว้ก่อน คลาสที่กำหนดให้เป็น `public` จะสามารถนำไปใช้งานนอกแพกเกจ (package) ได้ แต่ถ้าละตัวกำหนด `public` คลาสจะสามารถใช้งานได้เฉพาะภายในแพกเกจ เราจะกล่าวถึงเรื่องแพกเกจในภายหลัง
- `<field-declarations>` เป็นส่วนประกาศตัวแปรต่าง ๆ
- `<constructor-declarations>` เป็นส่วนประกาศตัวสร้าง (constructor) ซึ่งเป็นเมทอดชนิดพิเศษสำหรับกำหนดสถานะเริ่มต้นให้กับอ็อบเจกต์
- `<method-declarations>` เป็นส่วนประกาศเมทอดทั่วไป

ตัวแปรต่าง ๆ เมทอด และตัวสร้างสามารถประกาศภายในคลาสในลำดับใดก็ได้

**หมายเหตุ** เราไม่สามารถกำหนดการเข้าถึงของคลาสให้เป็นแบบ `private` หรือ `protected` ได้ ยกเว้นจะเป็นคลาสที่ซ้อนอยู่ภายในคลาสอื่น (nested class) อีกที

### การสร้างอ็อบเจกต์

การสร้างอ็อบเจกต์จะใช้คำสั่ง `new` ตามด้วยชื่อคลาสและอาร์กิวเมนต์ที่ต้องการส่งให้กับคอนสตรักเตอร์ของคลาส โดยมีรูปแบบวากยสัมพันธ์ดังนี้

```
new ClassName(<argument-list>)
```

โดยปกติแล้ว เมื่อเราสร้างอ็อบเจกต์ เราจะกำหนดตัวแปรเพื่อใช้อ้างอิงถึงตัวอ็อบเจกต์ที่สร้างขึ้นด้วย เราจึงมักเห็นการสร้างอ็อบเจกต์ในรูปของการประกาศตัวแปรไปพร้อมกับการสร้างดังนี้

```
ClassName varName = new ClassName(<argument-list>)
```

ในตัวอย่างของคลาส `Pair` เรามีส่วนของโปรแกรมที่สร้างอ็อบเจกต์ดังนี้

In [3]:
Pair p1 = new Pair();
Pair p2 = new Pair();

ในบางกรณี เราอาจจะสร้างอ็อบเจกต์ใหม่ได้โดยการเรียกเมทอดอื่นที่ทำหน้าที่สร้างอ็อบเจกต์ใหม่แล้วส่งคืนกลับมาให้ เมทอดประเภทนี้เรียกว่าเมทอดโรงงาน (factory method) เช่น ในตัวอย่างนี้

In [ ]:
LocalDateTime time = LocalDateTime.now();

เมทอด `now` จะทำหน้าที่สร้างอ็อบเจกต์ใหม่จากคลาส `LocalDateTime` โดยกำหนดให้มีค่าเป็นวันเวลาปัจจุบัน และส่งอ็อบเจกต์ที่สร้างขึ้นคืนให้เป็นผลลัพธ์

## โปรแกรมคือระบบนิเวศของอ็อบเจกต์

ม้าลายและนกกระจอกเทศมักจะชอบอยู่ใกล้ ๆ กัน ม้าลายมีสายตาที่เฉียบคม นกกระจอกเทศมีจมูกที่ไวต่อกลิ่น สัตว์ทั้งสองมีความสามารถต่างกัน แต่ทำงานร่วมกันโดยใช้ความสามารถของตนเพื่อระวังภัยจากนักล่ารูปแบบต่าง ๆ และส่งสัญญาณเตือนเมื่อพบเห็นภัยอันตราย เช่นเดียวกัน สุนัขป่าคาโยตี้กับแบดเจอร์ก็ทำงานร่วมกันเพื่อล่าหนูหรือกระรอกเป็นอาหาร โดยแบดเจอร์จะมุดดินเพื่อลากตัวหนูออกมาจากรู แล้วสุนัขป่าคาโยตี้ก็ใช้ความเร็วของมันในการไล่จับเหยื่อ ปรากฏการณ์เหล่านี้เป็นตัวอย่างของการทำงานร่วมกันของสิ่งต่าง ๆ โดยอาศัยความสามารถที่แตกต่างกันเพื่อบรรลุวัตถุประสงค์บางอย่าง

โปรแกรมที่เราเขียนขึ้นโดยใช้แนวคิดแบบ OOP นั้นจะประกอบด้วยอ็อบเจกต์ที่ทำงานร่วมกัน แต่ละอ็อบเจกต์อาจจะมาจากคลาสเดียวกันหรือต่างคลาสกัน และแต่ละคลาสอาจจะมีการสร้างเพียงอ็อบเจกต์เดียวหรือหลายอ็อบเจกต์ ตามแต่ลักษณะงานและความจำเป็น

เราอาจจะมีโปรแกรมที่ประกอบด้วยอ็อบเจกต์ต่าง ๆ จากคลาสตัวควบคุมเครื่องจ่ายน้ำ (`SprinklerController`) ตัวตั้งเวลา (`Timer`) ตัวอ่านข้อมูลจากตัวรับรู้ความชื้น (`HumiditySensorReader`) ตัวอ่านข้อมูลจากตัวรับรู้ความสว่าง (`LightSensorReader`) ตัวเก็บรวบรวมข้อมูล (`DataCollector`) และตารางเวลาการรดน้ำ (`WateringSchedule`) ทำงานประสานกันโดยการควบคุมและตัดสินใจของอ็อบเจกต์จากคลาสตัวควบคุมการรดน้ำพืช (`PlantWateringController`) ผ่านการตั้งค่าที่กำหนดโดยอ็อบเจกต์จากคลาสคอนโซลควบคุมระบบรดน้ำ (`WateringSystemConsole`)

ในกรณีตัวอย่างนี้ เราอาจต้องการใช้อ็อบเจกต์ `SprinklerController` หลายตัวเพื่อควบคุมเครื่องจ่ายน้ำหลายเครื่อง หรือ `HumiditySensorReader` และ `LightSensorReader` หลายตัวเพื่อติดต่อกับตัวรับรู้ (sensor) ในแต่ละจุด แต่อาจจะมีอ็อบเจกต์ `WateringSystemConsole` เป็นคอนโซลควบคุมเพียงตัวเดียวจัดการกับทั้งระบบ ทั้งนี้ขึ้นอยู่กับแนวคิดในการวิเคราะห์ปัญหาและการออกแบบระบบเพื่อแก้ไขปัญหา

### การตั้งชื่อแฟ้มข้อมูล

ในการเขียนโปรแกรมด้วยภาษา Java โดยปกติแล้วเราจะกำหนดให้มีคลาสเพียงคลาสเดียวต่อหนึ่งแฟ้มข้อมูล โดยเราจะตั้งชื่อแฟ้มข้อมูลเป็นชื่อเดียวกับชื่อของคลาส เช่น ในตัวอย่างข้างต้น คลาสชื่อว่า `Pair` เราจึงตั้งชื่อแฟ้มว่า `Pair.java` โปรแกรมหนึ่ง ๆ มักจะประกอบขึ้นจากหลาย ๆ คลาส จึงมักจะต้องประกอบด้วยแฟ้มข้อมูลมากมายตามจำนวนคลาสที่มีอยู่ในโปรแกรม

อย่างไรก็ดี ธรรมเนียมการมีคลาสเพียงคลาสเดียวต่อหนึ่งแฟ้มข้อมูล และการตั้งชื่อแฟ้มข้อมูลให้ตรงกับชื่อคลาสไม่ใช่ข้อบังคับที่จะต้องปฏิบัติทุกกรณี สิ่งที่เป็นข้อบังคับมีเพียง

1. ในหนึ่งแฟ้มข้อมูลอาจจะมีหลายคลาสได้ แต่จะมีคลาสที่เป็น `public` ได้เพียงคลาสเดียว
2. หากมีคลาสที่เป็น `public` แฟ้มข้อมูลนั้นจะต้องตั้งชื่อตามคลาสที่เป็น `public` นั้น แต่ถ้าหากไม่มีคลาสที่เป็น `public` เลย แฟ้มข้อมูลสามารถตั้งชื่ออย่างไรก็ได้

แนวปฏิบัติโดยทั่วไปนั้น เราไม่นิยมมีคลาสมากกว่าหนึ่งคลาสในหนึ่งแฟ้มข้อมูล ยกเว้นจะเป็นคลาสที่กำหนดขึ้นเพื่ออำนวยความสะดวกภายใน และไม่ต้องการใช้ในที่อื่น ๆ อีก ส่วนการตั้งชื่อแฟ้มข้อมูล ถึงแม้ในกรณีที่คลาสในแฟ้มข้อมูลนั้นจะไม่เป็น `public` โดยทั่วไปก็ยังนิยมที่จะตั้งชื่อแฟ้มข้อมูลตามชื่อคลาส

## เมทอด

อ็อบเจกต์แต่ละตัวมีคุณลักษณะของตัวเองในรูปของตัวแปรต่าง ๆ ซึ่งถูกกำหนดขึ้นและเปลี่ยนแปลงค่าได้ระหว่างการทำงานโดยส่วนของโค้ดที่เรียกว่าเมทอด (method) เมทอดเป็นตัวกำหนดพฤติกรรมของอ็อบเจกต์ว่าสามารถทำอะไรได้บ้างและทำอย่างไร ในตัวอย่างคลาส `Pair` เราจะเห็นเมทอด `getFirst`, `getSecond`, `setFirst`, `swap` และ `print` ซึ่งทำหน้าที่แตกต่างกันไปในนั้น เมทอด `getFirst` และ `getSecond` จะคืนค่าของตัวแปร `first` และ `second` กลับไปยังผู้เรียก `setFirst` จะกำหนดค่าให้กับตัวแปร `first` `swap` จะสลับค่าระหว่าง `first` และ `second` และ `print` จะแสดงค่าของ `first` และ `second` ออกมาทางจอภาพ

การกำหนดเมทอดจะทำในภายในคลาส โดยมีรูปแบบวากยสัมพันธ์เบื้องต้นดังนี้

```
<method-modifiers> <return-type> methodName(<parameter-list>)
    <method-body>
```

- `<method-modifiers>` เป็นตัวกำหนดคุณสมบัติของเมทอด ซึ่งอาจจะกำหนดระหว่าง `abstract` หรือเลือก `static`, `final` และ `synchronized` ประกอบกัน และสามารถมีตัวกำหนดการเข้าถึงได้ไม่เกินหนึ่งตัวจาก `private`, `protected` และ `public`
- `<return-type>` เป็นตัวระบุชนิดของข้อมูลที่จะส่งกลับจากเมทอด โดยที่ `void` จะหมายถึงเมทอดนี้ไม่ส่งค่ากลับ
- `<parameter-list>` เป็นรายการพารามิเตอร์ที่เมทอดนี้จะรับ
- `<method-body>` เป็นตัวโค้ดของเมทอดในรูปของบล็อก

## คอนสตรักเตอร์

คอนสตรักเตอร์ (constructor, ตัวสร้าง) เป็นเมทอดชนิดพิเศษที่มีหน้าที่กำหนดสถานะตั้งต้นให้กับอ็อบเจกต์ที่สร้างขึ้นใหม่แต่ละตัว โดยมีรูปแบบวากยสัมพันธ์เบื้องต้นดังนี้

```
<constructor-modifiers> ClassName(<parameter-list>)
    <constructor-body>
```

- `<constructor-modifiers>` เป็นตัวกำหนดคุณสมบัติของเมทอด ซึ่งอาจจะเป็น `private`, `protected`, `public` หรือละไว้ได้
- `<parameter-list>` เป็นรายการพารามิเตอร์ที่คอนสตรักเตอร์นี้จะรับ
- `<constructor-body>` เป็นตัวโค้ดของคอนสตรักเตอร์ในรูปของบล็อก

คอนสตรักเตอร์จะต้องมีชื่อเหมือนกับชื่อของคลาส และจะไม่มีการส่งค่ากลับเนื่องจากคอนสตรักเตอร์จะไม่ถูกเรียกใช้โดยตรง แต่จะถูกเรียกโดยอัตโนมัติเมื่อมีการสร้างอ็อบเจกต์ด้วยคำสั่ง `new`

In [1]:
import java.time.LocalDateTime;
import java.time.format.DateTimeFormatter;

public class Expense {
    private String item;
    private double amount;
    private LocalDateTime time;
    
    private static DateTimeFormatter formatter =
        DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm:ss");

    public Expense(String item, double amount) {
        this.item = item;
        this.amount = amount;
        time = LocalDateTime.now();
    }
    
    public String getItem() {
        return item;
    }
    
    public double getAmount() {
        return amount;
    }
    
    public void setAmount(double amount) {
        if (amount > 0.0) {
            this.amount = amount;
        }
    }
    
    public void print() {
        System.out.printf("(%s) %s: %.2f\n", time.format(formatter), item, amount);
    }
    
    public static void main(String[] args) {
        Expense item1 = new Expense("Phone bill", 799.00);
        item1.print();
    }
}

ในตัวอย่างนี้ คลาส `Expense` ซึ่งเราต้องการสร้างขึ้นเพื่อใช้จัดเก็บข้อมูลการใช้จ่ายหนึ่งรายการ มีคอนสตรักเตอร์ซึ่งมีชื่อเดียวกับคลาส รับพารามิเตอร์ 2 ตัว ได้แก่ `item` เป็น `String` แทนคำอธิบายรายการใช้จ่าย และ `amount` เป็น `double` แทนจำนวนเงินที่จ่ายไป คลาสนี้สามารถทำงานได้ด้วยตัวเองเนื่องจากมีเมทอด `main` อยู่ในคลาส

ใน `main` มีการสร้างอ็อบเจกต์ `Expense` ขึ้นหนึ่งตัว โดยให้ตัวแปร `item1` เป็นตัวอ้างอิง สังเกตว่าการเรียกคำสั่ง `new` เพื่อสร้างอ็อบเจกต์มีการส่งอาร์กิวเมนต์ไป 2 ตัว คือ "Phone bill" และ 799.00 ซึ่งอาร์กิวเมนต์ทั้ง 2 ตัวนี้จะถูกส่งไปเป็นพารามิเตอร์ของคอนสตรักเตอร์ของ `Expense`

คอนสตรักเตอร์ของ `Expense` ทำหน้าที่กำหนดสถานะตั้งต้นของอ็อบเจกต์ใหม่ โดยกำหนดให้ค่าของ `item` และ `amount` มีค่าตามที่ส่งเข้ามาและกำหนดให้ `time` มีค่าเป็นวันที่และเวลาปัจจุบัน คอนสตรักเตอร์จะถูกเรียกใช้งานโดยอัตโนมัติเฉพาะในตอนที่อ็อบเจกต์แต่ละตัวถูกสร้างขึ้นเท่านั้น เราไม่สามารถเรียกใช้ได้เองในภายหลังในตัวคอนตรักเตอร์นี้

## ตัวอ้างอิงอ็อบเจกต์ปัจจุบัน

เมทอดที่ไม่ได้เป็น `static` ทุกเมทอดและคอนสตรักเตอร์ เมื่อถูกเรียกใช้จะมีการส่งพารามิเตอร์แฝงมาด้วยตัวหนึ่งชื่อว่า `this`

`this` เป็นตัวแทนของอ็อบเจกต์ปัจจุบัน และสามารถใช้ในการอ้างอิงตัวแปร เมทอด และคอนสตรักเตอร์ของตัวอ็อบเจกต์เองได้ โดยทั่วไปแล้วเราจะเห็นการใช้ `this` ภายในตัวอ็อบเจกต์เองเวลาที่ชื่อที่ต้องการอ้างอิงถึงถูกบดบังโดยชื่ออื่น ในบางครั้งเราจะใช้ `this` เมื่ออ็อบเจกต์ต้องการส่งผ่านตัวมันเองไปยังเมทอดหรือคอนสตรักเตอร์ของอ็อบเจกต์อื่น นอกจากนี้ `this` ยังสามารถใช้โดยคอนสตรักเตอร์เพื่ออ้างอิงคอนตรักเตอร์ตัวอื่นได้อีกด้วย

ในคอนสตรักเตอร์ในตัวอย่าง เราจะเห็นตัวอ้างอิง `this` ในคำสั่ง `this.item = item` และ `this.amount = amount` ซึ่ง `this` จะหมายถึงตัวอ็อบเจกต์เอง `this.item = item` หมายถึง กำหนดค่า `item` ซึ่งเป็นพารามิเตอร์ของคอนตรักเตอร์ให้กับ `item` ซึ่งเป็นตัวแปรของอ็อบเจกต์ที่ประกาศเอาไว้ก่อนหน้านี้ `this.item` จะหมายถึง `item` ของตัวอ็อบเจกต์เอง ไม่ใช่ `item` ที่เป็นพารามิเตอร์ โดยมี `this` เป็นตัวระบุว่าเป็นตัวแปรของอ็อบเจกต์ ในกรณีตัวอย่างนี้ เราจำเป็นต้องระบุ `this` เนื่องจากพารามิเตอร์และตัวแปรของอ็อบเจกต์มีชื่อเหมือนกัน และ `item` ที่เป็นตัวแปรของอ็อบเจกต์ถูกบดบังโดย `item` ที่เป็นพารามิเตอร์ เราจึงต้องใช้ `this` ในการระบุให้ชัดเจนว่าเราหมายถึงตัวแปรของอ็อบเจกต์

ตัวแปร `time` ถูกอ้างอิงโดยตรงโดยไม่ต้องใช้ `this` เนื่องจากตัวแปร `time` ไม่ได้ถูกบดบังโดยตัวแปรอื่น เราจึงไม่จำเป็นต้องใช้ `this` เพื่อระบุให้เฉพาะเจาะจงลงไป

In [2]:
class Juggler {
    private int value;
    
    public Juggler(int value) {
        this.value = value;
    }
    
    public void swapWith(Juggler another) {
        int temp = this.value;
        this.value = another.value;
        another.value = temp;
    }
    
    public int getValue() {
        return value;
    }
}

เมื่อเราทดสอบได้ผลลัพธ์ดังนี้

In [3]:
Juggler a = new Juggler(10);
Juggler b = new Juggler(20);
a.swapWith(b);
System.out.printf("Juggler a = %d\n", a.getValue());
System.out.printf("Juggler b = %d\n", b.getValue());

Juggler a = 20
Juggler b = 10


ในตัวอย่างนี้ เราใช้ `this` ในคอนสตรักเตอร์และในเมทอด `swapWith` สังเกตว่าใน `swapWith` เราจะรับพารามิเตอร์ `another` เป็นอ็อบเจกต์ของคลาส `Juggler` อีกตัวหนึ่งด้วย คำสั่ง `this.value = another.value` จะเป็นการกำหนดค่าระหว่างอ็อบเจกต์ `Juggler` 2 ตัว คือตัวปัจจุบัน (ระบุโดย `this`) และตัวที่ถูกส่งผ่านเข้ามา (ระบุโดย `another`) ในกรณีเราไม่จำเป็นต้องระบุ `this` สำหรับอ็อบเจกต์ปัจจุบัน เนื่องจากตัวแปร `value` ไม่ถูกบดบัง แต่เราอาจจะเลือกที่จะระบุ `this` เพื่อความชัดเจนของโปรแกรมก็ได้

## ตัวกำหนดระดับการเข้าถึง

ในหัวข้อที่ผ่านมา เราจะเห็นตัวกำหนดระดับการเข้าถึงแบบ `public` และ `private` กำหนดให้กับคลาส เมทอด และตัวแปรต่าง ๆ ในภาษา Java จะมีการแบ่งระดับการเข้าถึงไว้ 4 ระดับ โดยเรียงลำดับจากที่เปิดมากที่สุดไปจนถึงจำกัดที่สุด ได้แก่ สาธารณะ (public), คุ้มครอง (protected), เฉพาะแพกเกจ (package-private) และส่วนตัว (private) โดยใช้ตัวกำหนด `public`, `protected` และ `private` ส่วนระดับการเข้าถึงแบบเฉพาะแพกเกจจะไม่มีตัวกำหนด เนื่องจากเป็นระดับการเข้าถึงโดยปริยายซึ่งถูกกำหนดโดยอัตโนมัติถ้าเราไม่กำหนดให้เป็นแบบอื่น

การกำหนดระดับการเข้าถึงแบบต่าง ๆ ให้กับสมาชิกของคลาส ได้แก่ ตัวแปร เมทอด คลาสซ้อน และอินเทอร์เฟซ มีความหมายโดยเรียงตามระดับความจำกัดดังนี้

- ระดับสาธารณะ ใช้ตัวกำหนด `public` สมาชิกที่ถูกกำหนดเป็นสาธารณะจะสามารถอ้างอิงได้จากทุกคลาสในแพกเกจใดก็ได้
- ระดับคุ้มครอง ใช้ตัวกำหนด `protected` สมาชิกที่ถูกกำหนดเป็นระดับคุ้มครองจะสามารถอ้างอิงได้จากคลาสภายในแพกเกจเดียวกัน และจากภายในตัวอ็อบเจกต์จากคลาสที่สืบทอดจากคลาสที่ประกาศสมาชิกนี้
- ระดับเฉพาะแพกเกจ ไม่ต้องระบุตัวกำหนด สมาชิกที่ถูกกำหนดเป็นระดับเฉพาะแพกเกจจะสามารถอ้างอิงได้จากคลาสที่อยู่ภายในแพกเกจเดียวกันเท่านั้น
- ระดับส่วนตัว ใช้ตัวกำหนด `private` สมาชิกที่ถูกกำหนดให้เป็นส่วนตัวจะสามารถอ้างอิงได้เฉพาะภายในคลาสที่ประกาศสมาชิกนี้และคลาสที่ซ้อนอยู่ในคลาสนี้เท่านั้น

ในกรณีของการกำหนดระดับการเข้าถึงให้กับตัวคลาสเอง ความหมายจะต่างไปจากกรณีของตัวแปร เมทอด หรือสมาชิกอื่น ๆ ของคลาส คลาสทั่วไปที่ไม่ได้ซ้อนอยู่ในคลาสอื่นจะมีระดับการเข้าถึงได้เพียงแบบสาธารณะและแบบเฉพาะแพกเกจเท่านั้น เนื่องจากระดับการเข้าถึงแบบคุ้มครองและแบบส่วนตัวจะมีความหมายเฉพาะกรณีของการอ้างอิงตัวแปรหรือเมทอด ถ้าคลาสถูกกำหนดให้เป็นส่วนตัวได้ จะไม่มีใครสามารถใช้งานคลาสนั้นได้เลย และเนื่องจากแพกเกจไม่สามารถสืบทอดได้ การกำหนดระดับการเข้าถึงแบบคุ้มครองจึงไม่มีความหมายสำหรับคลาส จึงไม่มีประโยชน์ที่จะอนุญาตให้กำหนดแบบนั้นได้

ในกรณีที่เป็นคลาสที่ซ้อนอยู่ในคลาสอื่นเราจะสามารถกำหนดระดับการเข้าถึงแบบคุ้มครองและและแบบส่วนตัวได้ เพราะการกำหนดให้เป็นแบบส่วนตัวหรือแบบคุ้มครองในระดับนี้หมายถึงให้อ้างอิงได้ภายในคลาสที่ครอบมันอยู่ (และอ็อบเจกต์จากคลาสที่สืบทอดมาจากคลาสที่ครอบอยู่ในกรณีแบบคุ้มครอง) ซึ่งมีประโยชน์ในการใช้งานจริง

เราจะกล่าวถึงระดับการเข้าถึงแบบคุ้มครองโดยละเอียดอีกครั้งเมื่อเราพูดถึงคุณสมบัติการสืบทอด (inheritance) ของ OOP

## ตัวเปลี่ยนแปลงค่าและตัวเข้าถึง

ในแนวคิดแบบ OOP เรามักจะพบเห็นเมทอดที่ทำหน้าที่เฉพาะคือการเปลี่ยนแปลงค่าหรือการอ่านค่าของตัวแปรต่าง ๆ เช่น ในตัวอย่างของคลาส `Expense` เราจะมีเมทอด `getItem` และ `getAmount` ที่ใช้อ่านค่าตัวแปร `item` และ `amount` และเมทอด `setAmount` สำหรับกำหนดค่าใหม่ให้กับตัวแปร `amount`

เมทอดสำหรับอ่านค่า เราเรียกว่าตัวเข้าถึง (accessor) หรือตัวอ่านค่า (getter) ธรรมเนียมของ Java จะตั้งชื่อเมทอดเหล่านี้โดยขึ้นต้นด้วยคำว่า `get` และเมทอดสำหรับเปลี่ยนแปลงค่า เราเรียกว่าตัวเปลี่ยนแปลงค่า (mutator) หรือตัวตั้งค่า (setter) โดยธรรมเนียมเราจะตั้งชื่อเมทอดกลุ่มนี้โดยขึ้นต้นด้วยคำว่า `set`

มาถึงตรงนี้อาจจะเกิดคำถามว่า ทำไมเราจึงต้องมีเมทอดเหล่านี้ ทำไมเราไม่ให้ผู้ใช้เข้าถึงและเปลี่ยนแปลงค่าของตัวแปรต่าง ๆ ในอ็อบเจกต์โดยตรง

ในสภาพแวดล้อมการพัฒนาโปรแกรมจริง ซึ่งอาจจะมีผู้พัฒนาร่วมกันหลายคน หรือเป็นการพัฒนาโปรแกรมหรือไลบรารีที่จะมีคนนำไปพัฒนาต่อ หรือแม้แต่โปรแกรมที่พัฒนาคนเดียวแต่มีความซับซ้อนถึงระดับหนึ่ง การจัดการการเข้าถึงและเปลี่ยนแปลงค่าต่าง ๆ ของอ็อบเจกต์ที่กำลังทำงานอยู่เป็นสิ่งที่สำคัญมาก ปัญหาบางประการที่เกี่ยวข้องกับการอนุญาตให้เข้าถึงและเปลี่ยนแปลงค่าภายในของอ็อบเจกต์ได้โดยตรงได้แก่

- อาจเกิดการเปลี่ยนแปลงค่าในรูปแบบที่เราไม่อนุญาต เช่น มีการตั้งค่าเป็นลบให้กับปริมาณที่เราไม่อนุญาตให้มีค่าติดลบได้
- ขั้นตอนหรือวิธีการเข้าถึงข้อมูลอาจไม่เป็นไปตามที่เรากำหนด เราอาจจะมีการกำหนดลำดับการทำงานเพื่อการเข้าถึงหรือเปลี่ยนแปลงข้อมูลเอาไว้ แต่ผู้พัฒนาคนอื่นหรือตัวเราเองอาจจะลืมที่จะทำตามขั้นตอนเหล่านั้น
- เกิดการเปลี่ยนแปลงค่าของอ็อบเจกต์ในเวลาที่ไม่ได้คาดเอาไว้ และการติดตามหาตำแหน่งของโปรแกรมที่ทำให้เกิดการเปลี่ยนแปลงค่านั้นมักจะเป็นเรื่องยาก
- โปรแกรมส่วนอื่นไปขึ้นตรงกับตัวแปรภายในอ็อบเจกต์นี้ ทำให้การเปลี่ยนแปลงการออกแบบหรือรูปแบบการแทนข้อมูลภายในอ็อบเจกต์ทำได้ยาก เนื่องจากจะไปกระทบกับโปรแกรมส่วนอื่นที่ขึ้นตรงอยู่

แนวทางการแก้ไขปัญหานี้โดยทั่วไปจะจัดการดังนี้

1. กำหนดให้ตัวแปรของอ็อบเจกต์ทุกตัวมีระดับการเข้าถึงเป็นแบบส่วนตัว ซึ่งจะทำให้คลาสอื่น ๆ ไม่สามารถเข้าถึงตัวแปรเหล่านี้ได้โดยตรง
2. กำหนดตัวเข้าถึงให้กับทุกค่าที่ต้องการให้ภายนอกสามารถอ่านค่าได้ เราสามารถเลือกได้ว่าต้องการให้สามารถเข้าถึงค่าใดได้บ้าง โดยค่าที่ไม่มีตัวเข้าถึงก็จะไม่สามารถอ่านได้ เพราะเราได้กำหนดตัวแปรต่าง ๆ ให้มีระดับการเข้าถึงเป็นส่วนตัวแล้ว
3. กำหนดตัวเปลี่ยนแปลงค่าให้กับทุกค่าที่ต้องการให้ภายนอกสามารถเปลี่ยนแปลงค่าได้ ในตัวเปลี่ยนแปลงค่าเราสามารถกำหนดเงื่อนไขและขั้นตอนในการเปลี่ยนแปลงค่าได้ ในตัวอย่างของคลาส `Expense` เมทอด `setAmount` เป็นตัวเปลี่ยนแปลงค่าซึ่งมีการตรวจสอบค่าที่ส่งเข้ามาก่อนว่าเป็นค่าติดลบหรือไม่ ถ้าไม่ติดลบก็จะเปลี่ยนค่าตัวแปร `amount` ให้เป็นไปตามที่ส่งเข้ามา แต่ถ้าติดลบก็จะผ่านไปเฉย ๆ โดยไม่เปลี่ยนแปลงค่าใด ๆ การใช้ตัวเปลี่ยนแปลงค่าทำให้เราสามารถบังคับรูปแบบการเปลี่ยนค่าของตัวแปรต่าง ๆ ให้เป็นไปตามที่เราต้องการได้ ตัวแปรใดที่เราไม่ต้องการให้เปลี่ยนแปลงค่าได้เลยก็จะไม่มีตัวเปลี่ยนแปลงค่า และเนื่องจากตัวแปรเหล่านั้นมีระดับการเข้าถึงเป็นแบบส่วนตัว ก็จะไม่มีการเข้าไปเปลี่ยนค่าจากภายนอกในทางอื่น ๆ ได้

การจัดการตามวิธีการนี้จะแก้ปัญหาที่กล่าวมาข้างต้นได้ เนื่องจากการเข้าถึงค่าต่าง ๆ ในอ็อบเจกต์จะถูกควบคุมโดยเมทอดของเราทั้งหมด และจะไม่มีโค้ดภายนอกคลาสที่จะมาขึ้นอยู่กับตัวแปรหรือรูปแบบของค่าภายในคลาสได้ เนื่องจากเราปิดการเข้าถึงโดยตรงเอาไว้ด้วยการกำหนดให้เป็นส่วนตัว

> **หมายเหตุ** ในตัวอย่างคลาส `Expense` เมทอด `print` ไม่ถือเป็นตัวเปลี่ยนแปลงค่าหรือตัวเข้าถึง เพราะ `print` ไม่ได้คืนค่าและไม่ได้เปลี่ยนแปลงค่าใด ๆ ภายในอ็อบเจกต์

### ธรรมเนียมการตั้งชื่อตัวเปลี่ยนแปลงค่าและตัวเข้าถึง

เราอาจจะตั้งชื่อเมทอดที่เป็นตัวเปลี่ยนแปลงค่าหรือตัวเข้าถึงอย่างไรก็ได้ แต่โดยธรรมเนียมแล้วตัวเปลี่ยนแปลงค่าที่ทำหน้าที่กำหนดค่าโดยตรงมักจะถูกตั้งชื่อโดยขึ้นต้นด้วยคำว่า `set` เช่น `setSalary` และตัวเข้าถึงมักจะถูกตั้งชื่อขึ้นต้นด้วยคำว่า `get` เช่น `getAmount`

ตัวเข้าถึงและตัวเปลี่ยนแปลงค่าอาจจะไม่ต้องกระทำกับตัวแปรชื่อเดียวกันโดยตรงเสมอไป ในทางปฏิบัติแล้วเราสามารถที่จะกำหนดการแทนข้อมูลภายในอ็อบเจกต์อย่างไรก็ได้ ตัวเข้าถึงและตัวเปลี่ยนแปลงค่าเป็นเพียงหน้าฉากที่เราต้องการให้ผู้เรียกใช้เห็น เช่น เราอาจจะมีตัวแปร `day`, `month` และ `year` แทนข้อมูลวันที่ แต่เราอาจจะมีเมทอด `setDate(day, month, year)` เมทอดเดียวเป็นตัวเปลี่ยนแปลงค่าสำหรับทั้ง 3 ตัวแปรนั้นเลยโดยไม่ต้องมีตัวแปรชื่อ `date` นอกจากนี้ เราอาจจะมีเมทอด `getDay`, `getMonth` และ `getYear` เป็นตัวเข้าถึงแยกเป็นรายค่าได้ ไม่จำเป็นต้องสอดคล้องกับตัวเปลี่ยนแปลงค่า ในทางกลับกัน เราอาจจะมีเมทอด `getTotalAmount` ซึ่งเป็นผลรวมของค่าทั้งหมด ไม่ได้มาจากตัวแปรตัวเดียว การเลือกออกแบบแบบใดเป็นสิ่งที่ต้องพิจารณาจากการใช้งานจริงเป็นหลัก

เมทอดที่เปลี่ยนแปลงค่าแต่เปลี่ยนแปลงในรูปแบบอื่น ๆ ที่ไม่ได้เป็นการกำหนดค่าให้ใหม่ ควรตั้งชื่อตามความเหมาะสม เช่น `addExpense`, `raiseSalary`, `removeItem` หรือ `sort` เป็นต้น

เมทอดที่ใช้ทดสอบสถานะของอ็อบเจกต์ เช่น ทดสอบว่าเครื่องจ่ายน้ำกำลังทำงานอยู่หรือหยุดทำงาน หรือทดสอบว่าปีนั้นเดือนกุมภาพันธ์มี 28 วันหรือ 29 วัน จัดเป็นตัวเข้าถึง มีการส่งค่ากลับเป็น `boolean` แสดงผลการทดสอบว่าจริงหรือเท็จ ควรตั้งชื่อขึ้นต้นด้วยคำว่า `is` เช่น `isWorking`, `isLeapYear` หรือ `isEnabled` เป็นต้น

## ตัวแปรของอินสแตนซ์

ตัวแปรที่เราประกาศขึ้นสำหรับอ็อบเจกต์เรียกว่าตัวแปรของอินสแตนซ์ (instance variable) ซึ่งตัวแปรเหล่านี้จะถูกสร้างขึ้นพร้อมกับตัวอ็อบเจกต์เมื่อเราสร้างอ็อบเจกต์ใหม่ ตัวแปรที่ประกาศในคลาสโดยไม่มีตัวกำหนด `static` ถือเป็นตัวแปรของอินสแตนซ์

## ตัวแปรของคลาส

ตัวแปรที่ประกาศในคลาสโดยมีตัวกำหนด `static` จะถือเป็นตัวแปรของคลาส (class variable) หรือตัวแปรสถิต (static variable) และไม่มีอ็อบเจกต์ตัวใดเป็นเจ้าของ พิจารณาตัวอย่างต่อไปนี้

In [1]:
public class Employee {
    private String name;
    private double salary;
    private final int id;
    private static int lastId = 1000;
    
    public static final double SALARY_STEP_SIZE = 10.0;
    
    public Employee(String name, double salary) {
        this.name = name;
        this.salary = computeNextSalaryStep(salary);
        id = ++lastId;
    }

    public static double computeNextSalaryStep(double salary) {
        // Round to the next salary step
        double steps = Math.ceil(salary / SALARY_STEP_SIZE);
        return steps * SALARY_STEP_SIZE;
    }
    
    public void raiseSalary(double percent) {
        double raise = salary * percent / 100.0;
        salary = computeNextSalaryStep(salary + raise);
    }
    
    public void showProfile() {
        System.out.printf("Name: %s%n", name);
        System.out.printf("ID: %d%n", id);
        System.out.printf("Salary: %,.2f%n", salary);
    }
}

คลาส Employee มีการประกาศตัวแปรทั้งหมด 5 ตัว โดยมี `name`, `salary` และ `id` เป็นตัวแปรของอินสแตนซ์ และมี `lastId` กับ `SALARY_STEP_SIZE` เป็นตัวแปรของคลาสเนื่องจากมีการระบุตัวกำหนด `static`

พฤติกรรมของตัวแปรของคลาสสามารถเห็นได้จากตัวอย่างการใช้งานในส่วนของโปรแกรมนี้

In [13]:
Employee george = new Employee("George", 15_233.00);
Employee sarah = new Employee("Sarah", 18_500.00);

george.showProfile();

sarah.raiseSalary(12.5);
sarah.showProfile();

Name: George
ID: 1001
Salary: 15,240.00
Name: Sarah
ID: 1002
Salary: 20,820.00


อ็อบเจกต์ `Employee` ถูกสร้างขึ้น 2 ตัวและอ้างอิงด้วยตัวแปร `george` และ `sarah` อ็อบเจกต์ทั้ง 2 ตัวนี้จะมีค่าของ `name`, `salary` และ `id` ของตัวเองที่แตกต่างกัน

คอนสตรักเตอร์ของ `Employee` มีการกำหนดค่ารหัสประจำตัวพนักงาน (`id`) ให้อ็อบเจกต์ที่สร้างใหม่ทุกอ็อบเจกต์โดยอัตโนมัติด้วยคำสั่ง `id = ++lastId` ซึ่งจะเป็นการวิ่งเลขตั้งแต่ 1001 ขึ้นไปเรื่อย ๆ ตามลำดับของอ็อบเจกต์ที่ถูกสร้างขึ้น

ตัวแปรที่สำคัญในงานนี้คือ `lastId` ซึ่งประกาศเป็น `static` และมีค่าเริ่มต้นที่ 1000 ตัวแปรนี้เป็นตัวแปรของคลาส ซึ่งในกรณีนี้หมายความว่า `lastId` เป็นของคลาส `Employee` ไม่ได้อยู่ที่อ็อบเจกต์ตัวใดตัวหนึ่ง ไม่ได้อยู่ที่ `george` และไม่ได้อยู่ที่ `sarah`

และเมื่อ `lastId` อยู่ที่คลาส `Employee` ก็หมายความว่าเรามีตัวแปร `lastId` อยู่เพียงตัวเดียว ทุกอ็อบเจกต์ที่เข้าถึงตัวแปรนี้จะเข้าถึงตัวแปรตัวเดียวกัน คำสั่ง `id = ++lastId` จะหมายถึง เพิ่มค่าให้กับตัวแปร `lastId` ซึ่งมีอยู่เพียงตัวเดียว แล้วกำหนดค่านั้นให้กับตัวแปร `id` ซึ่งแต่ละอ็อบเจกต์จะมีเป็นของตัวเอง

เนื่องจาก `lastId` มีอยู่ตัวเดียว และเราเพิ่มค่าให้เรื่อย ๆ ทุกครั้งที่มีการสร้างอ็อบเจกต์จากคลาส `Employee` ใหม่ ตัวแปร `id` ของอ็อบเจกต์ใหม่ก็จะได้ค่าใหม่จาก `lastId` ตัวเดียวกันซึ่งเพิ่มค่าขึ้นไปเรื่อย ๆ ทำให้อ็อบเจกต์แต่ละตัวมีค่า `id` ที่ไม่ซ้ำกัน

การอ้างอิงตัวแปรจากภายนอกคลาส ตัวแปรที่เป็น `static` สามารถอ้างอิงผ่านคลาสได้เลยโดยไม่ต้องมีการสร้างอ็อบเจกต์ขึ้นมาก่อน แต่ตัวแปรของอินสแตนซ์จะต้องอ้างอิงผ่านอ็อบเจกต์เสมอ

In [ ]:
System.out.println(Employee.lastId); // OK
System.out.println(george.lastId); // OK
System.out.println(Employee.id); // ERROR: referencing non-static variable from static context

จะเห็นว่า `lastId` ซึ่งเป็น `static` จะสามารถอ้างอิงผ่านชื่อคลาส `Employee` ก็ได้ หรือจะอ้างอิงผ่านอ็อบเจกต์ก็ได้ และจะหมายถึงตัวแปรตัวเดียวกัน แต่ `id` ซึ่งเป็นตัวแปรของอินสแตนซ์จะไม่สามารถอ้างอิงผ่านคลาสได้ ทั้งนี้เพราะว่า `id` ผูกอยู่กับตัวอ็อบเจกต์ เราจึงไม่สามารถอ้างขึ้นมาลอย ๆ โดยไม่ระบุว่าเป็น `id` ของอ็อบเจกต์ตัวไหนได้

### ตัวกำหนด final กับตัวแปร

ตัวแปร `id` และ `SALARY_STEP_SIZE` มีการประกาศด้วยตัวกำหนด `final` ด้วย ซึ่งจะหมายถึงว่าเมื่อกำหนดค่าให้ตัวแปรเหล่านี้แล้ว ค่านั้นจะเปลี่ยนแปลงอีกไม่ได้ โดยการกำหนดค่าให้กับตัวแปรที่เป็น final จะต้องทำตอนประกาศตัวแปร ในคอนสตรักเตอร์ หรือในบล็อกกำหนดค่าเริ่มต้น (initializer block) เท่านั้น

ตัวแปร `id` ต่างจาก `SALARY_STEP_SIZE` อีกจุดหนึ่งคือ `id` เป็นตัวแปรของอินสแตนซ์ แต่ `SALARY_STEP_SIZE` เป็นตัวแปรของคลาส (ประกาศด้วย `static`) ถึงแม้ตัวแปร `id` จะไม่เปลี่ยนแปลงค่า แต่ก็จะมีค่าแตกต่างกันไปในแต่ละอ็อบเจกต์ แต่ `SALARY_STEP_SIZE` จะมีค่าเพียงค่าเดียวไม่เปลี่ยนแปลง ไม่ได้มีค่าแตกต่างกันไปในแต่ละอ็อบเจกต์ เพราะเป็นตัวแปรของคลาส

ตัวแปรของคลาสที่ไม่เปลี่ยนค่านี้เราเรียกอีกอย่างว่าค่าคงที่ (constant) เนื่องจากค่าของตัวแปรเหล่านั้นมีค่าเดียวและจะคงค่านั้นไปตลอดการทำงาน เพราะฉะนั้นค่าคงที่จะต้องประกาศด้วยคำว่า `static final`

การตั้งชื่อค่าคงที่ใน Java จะมีธรรมเนียมให้ใช้อักษรตัวพิมพ์ใหญ่ทั้งหมด และแบ่งคำด้วย `_` (underscore) อย่างในชื่อของ `SALARY_STEP_SIZE` ส่วน `id` ไม่จัดเป็นค่าคงที่ถึงแม้ค่าจะไม่เปลี่ยนแปลงก็ตาม เพราะมีค่าแตกต่างกันไปตามแต่ละอ็อบเจกต์ได้ จึงไม่ได้ใช้ธรรมเนียมนี้

> **หมายเหตุ** การประกาศให้ตัวแปรที่เป็นตัวอ้างอิงอ็อบเจกต์หรืออาร์เรย์เป็นค่าคงที่ไม่ได้หมายความว่าอ็อบเจกต์นั้นจะเปลี่ยนค่าไม่ได้ แต่หมายความว่าตัวแปรที่อ้างอิงอ็อบเจกต์นั้นจะเปลี่ยนไปอ้างอิงอ็อบเจกต์อื่นอีกไม่ได้ ตัวอ็อบเจกต์หรืออาร์เรย์เองสามารถเปลี่ยนแปลงค่าได้ตามปกติ

## เมทอดของคลาส

ในตัวอย่างคลาส `Employee` มีเมทอดหนึ่งที่ประกาศด้วยตัวกำหนด `static` ได้แก่เมทอด `computeNextSalaryStep` ซึ่งใช้คำนวณการปัดเงินเดือนขึ้นไปยังขั้นเงินเดือนถัดไปหลังการปรับเงินเดือน

เมทอดที่ประกาศด้วย `static` เราเรียกว่าเมทอดของคลาส (class method) หรือเมทอดสถิต (static method) เราสามารถเรียกใช้เมทอดของคลาสผ่านคลาสได้โดยตรงโดยไม่ต้องสร้างอ็อบเจกต์ก่อน ในกรณีของ `Employee` เราสามารถเรียก `Employee.computeNextSalaryStep(15_233.00)` ได้โดยตรงโดยไม่ต้องมีอ็อบเจกต์ใด ๆ

อย่างไรก็ดี เมทอดของคลาสมีข้อจำกัดที่สำคัญอันหนึ่งคือ เมทอดของคลาสจะเข้าถึงได้เฉพาะตัวแปรของคลาสเท่านั้น ไม่สามารถเข้าถึงตัวแปรของอินสแตนซ์ได้โดยตรง เพราะว่าเมทอดของคลาสไม่ได้ผูกอยู่กับอ็อบเจกต์ใด ๆ จึงไม่มีตัวแปรของอินสแตนซ์ให้อ้างอิงได้ นอกจากในกรณีที่มีการส่งตัวอ้างอิงของอ็อบเจกต์มาให้เป็นอาร์กิวเมนต์ ในกรณีนั้นก็จะสามารถอ้างอิงตัวแปรของอินสแตนซ์ผ่านตัวอ้างอิงของอ็อบเจกต์นั้น ๆ ได้

> **หมายเหตุ** เมทอดที่ไม่ได้กำหนดเป็น `static` เรียกว่าเมทอดของอินสแตนซ์ (instance method) ซึ่งผูกอยู่กับอ็อบเจกต์ ซึ่งในตัวอย่าง `Employee` ก็คือทุกเมทอดยกเว้น `computeNextSalaryStep` การเรียกใช้เมทอดประเภทนี้ต้องเรียกผ่านอ็อบเจกต์ และตัวอ็อบเจกต์ที่ถูกใช้อ้างอิงจะส่งผ่านตัวเองเข้าไปในเมทอดด้วยในรูปของ `this` ซึ่งเราสามารถนำไปใช้อ้างอิงในเมทอดได้

## การโอเวอร์โหลด

กลับมาที่ตัวอย่าง `Pair` อีกครั้ง เราปรับปรุง `Pair` โดยการเพิ่มคอนสตรักเตอร์เพื่อกำหนดค่าเริ่มต้นให้กับอ็อบเจกต์ใหม่ของ `Pair` และเพิ่มให้สามารถสลับค่าได้ทั้งภายในและสลับกับอ็อบเจกต์ของ `Pair` อีกตัว

In [5]:
public class Pair {
    private int first;
    private int second;
    
    // Parameter-less constructor -- default to zeros
    public Pair() {
        this(0, 0);
    }
    
    // Two-parameter constructor
    public Pair(int first, int second) {
        this.first = first;
        this.second = second;
    }
    
    public int getFirst() {
        return first;
    }
    
    public int getSecond() {
        return second;
    }
    
    // Swap between first and second
    public void swap() {
        int temp = first;
        first = second;
        second = temp;
    }
    
    // Swap between 2 pairs
    public void swap(Pair another) {
        int temp = first;
        first = another.first;
        another.first = temp;
        
        temp = second;
        second = another.second;
        another.second = temp;
    }
    
    public void print() {
        System.out.println("(" + first + ", " + second + ")");
    }
}

ในตัวอย่างนี้เราจะพบว่ามีเมทอดชื่อ `swap` 2 เมทอด และมีคอนสตรักเตอร์ 2 ตัวด้วยเช่นกัน

การสร้างเมทอดหรือคอนสตรักเตอร์ที่ชื่อเหมือนกับเมทอดที่มีอยู่แล้วโดยที่มีจำนวนหรือชนิดของพารามิเตอร์ต่างกัน เราเรียกว่าการโอเวอร์โหลดเมทอด (method overloading) หรือการโอเวอร์โหลดคอนสตรักเตอร์ (constructor overloading)

เวลาที่เราเรียกเมทอดที่มีการโอเวอร์โหลด คอมไพเลอร์จะเลือกเมทอดที่เหมาะสมจากอาร์กิวเมนต์ที่ส่งผ่านเข้าไป โดยดูว่าจำนวนอาร์กิวเมนต์เท่ากับจำนวนพารามิเตอร์ของเมทอด และชนิดของอาร์กิวเมนต์เข้ากันได้กับชนิดของพารามิเตอร์มากที่สุด เช่นเดียวกับคอนสตรักเตอร์ เมื่อเราสร้างอ็อบเจกต์ด้วยคำสั่ง `new` คอมไพเลอร์จะเลือกคอนสตรักเตอร์ที่เหมาะสมจากอาร์กิวเมนต์ที่ผ่านเข้าไปเช่นกัน

เราลองทดสอบการสร้างและการเรียกเมทอดของ `Pair` ดู

In [6]:
Pair p = new Pair(1, 2);
Pair q = new Pair();

p.print(); // (1, 2)
q.print(); // (0, 0)

p.swap(); // p.first <-> p.second
p.print(); // (2, 1)

p.swap(q); // p.first <-> q.first, p.second <-> q.second
p.print(); // (0, 0)
q.print(); // (2, 1)

(1, 2)
(0, 0)
(2, 1)
(0, 0)
(2, 1)


ในส่วนแรก การสร้างอ็อบเจกต์ `p` เราใช้คอนสตรักเตอร์ที่รับพารามิเตอร์ 2 ตัวแล้วกำหนดค่าให้ `first` กับ `second` แต่การสร้างอ็อบเจกต์ `q` เราใช้คอนสตรักเตอร์ที่ไม่มีพารามิเตอร์

คอนสตรักเตอร์ที่โอเวอร์โหลดสามารถเรียกกันเองได้โดยใช้ `this` เป็นตัวแทนคอนสตรักเตอร์ ในตัวอย่างนี้คอนสตรักเตอร์ตัวที่ไม่มีพารามิเตอร์เรียกใช้คอนสตรักเตอร์ตัวที่มี 2 พารามิเตอร์ด้วยผ่าน `this` โดยส่งอาร์กิวเมนต์ไปเป็น 0 ทั้ง 2 ค่า

ตอนที่เราเรียก `p.swap()` เมทอด `swap` ที่ถูกเรียกก็คือ `swap` ที่ไม่มีพารามิเตอร์ และการทำงานก็คือการสลับค่าระหว่าง `first` และ `second` ภายใน `p` แต่เมื่อเราเรียก `p.swap(q)` เมทอด `swap` ที่ถูกเรียกคือ `swap` ที่รับพารามิเตอร์เป็นอ็อบเจกต์ `Pair` อีกตัวหนึ่งแล้วสลับค่าของ `first` และ `second` กับอ็อบเจ็กต์ที่ถูกส่งผ่านเข้ามา

การเลือกเมทอดที่จะถูกเรียกมีรายละเอียดที่น่าสนใจเมื่อพิจารณาประกอบกับการแปลงชนิดโดยอัตโนมัติของ Java

ลองพิจารณาตัวอย่างต่อไปนี้

In [23]:
public class OverloadingTest {
    // methodA: Direct type tests
    public void methodA(int v) {
        System.out.println("The int version of methodA is called with value: " + v);
    }

    public void methodA(long v) {
        System.out.println("The long version of methodA is called with value: " + v);
    }

    public void methodA(double v) {
        System.out.println("The double version of methodA is called with value: " + v);
    }

    public void methodA(Integer v) {
        System.out.println("The Integer version of methodA is called with value: " + v);
    }
    
    // methodB: Type widening tests
    public void methodB(int v) {
        System.out.println("The int version of methodB is called with value: " + v);
    }

    public void methodB(double v) {
        System.out.println("The double version of methodB is called with value: " + v);
    }
    
    // methodC: Autoboxing tests
    public void methodC(Float v) {
        System.out.println("The Float version of methodC is called with value: " + v);
    }

    public void methodC(double v) {
        System.out.println("The double version of methodC is called with value: " + v);
    }
    
    // methodD: Choosing between widening and autoboxing
    public void methodD(long v) {
        System.out.println("The long version of methodD is called with value: " + v);
    }

    public void methodD(Integer v) {
        System.out.println("The Integer version of methodD is called with value: " + v);
    }
}

OverloadingTest c = new OverloadingTest();

System.out.println("methodA: Direct type tests");
c.methodA(101);
c.methodA(102L);
c.methodA(104.0);
c.methodA(new Integer(105));

System.out.println("methodB: Type widening tests");
c.methodB(102L);
c.methodB(103.0f);

System.out.println("methodC: Autoboxing tests");
c.methodC(103.0f);
c.methodC(new Double(106.0));

System.out.println("methodD: Choosing between widening and autoboxing");
c.methodD(101);

methodA: Direct type tests
The int version of methodA is called with value: 101
The long version of methodA is called with value: 102
The double version of methodA is called with value: 104.0
The Integer version of methodA is called with value: 105
methodB: Type widening tests
The double version of methodB is called with value: 102.0
The double version of methodB is called with value: 103.0
methodC: Autoboxing tests
The double version of methodC is called with value: 103.0
The double version of methodC is called with value: 106.0
methodD: Choosing between widening and autoboxing
The long version of methodD is called with value: 101


การทดสอบด้วย `methodA` แสดงให้เห็นว่าถ้ามีเมทอดที่มีชนิดของพารามิเตอร์ตรงกับชนิดของอาร์กิวเมนต์ เมทอดนั้นจะถูกเลือกทันที ส่วนในกรณีที่ชนิดไม่ตรงกัน เราทดสอบด้วย `methodB`, `methodC` และ `methodD` ซึ่งได้ผลลัพธ์ดังนี้

`methodB` แสดงให้เห็นว่าถ้ามีเมทอดที่พารามิเตอร์มีชนิดที่กว้างกว่าอาร์กิวเมนต์ เมทอดนั้นก็จะถูกเลือก ในตัวอย่างนี้เราจะเห็นว่าเมื่อส่งอาร์กิวเมนต์เป็นชนิด `long` เมทอดที่รับ `double` ก็ถูกเรียก และถ้าส่งเป็น `float` เมทอดที่รับ `double` ก็ยังถูกเรียก เพราะว่าอีกเมทอดรับ `int` ซึ่งแคบกว่า (หรือเล็กกว่า) `long` และ `float` การแปลงชนิดโดยอัตโนมัติเพื่อส่งผ่านเป็นอาร์กิวเมนต์จะเป็นการแปลงให้กว้างขึ้น (type widening) เท่านั้น

กรณีที่อาร์กิวเมนต์หรือพารามิเตอร์เป็นคลาสหีบห่อ (wrapper class) Java จะแปลงไปเป็นคลาสหีบห่อไปหรือแปลงกลับให้โดยอัตโนมัติ (autoboxing) ตัวอย่าง `methodC` แสดงให้เห็นถึงการแปลงจาก `float` ไปเป็น `Float` และจาก `Double` กลับมาเป็น `double` โดยอัตโนมัติ

ในกรณีที่ต้องเลือกระหว่างการแปลงชนิดให้กว้างขึ้นกับการแปลงเป็นคลาสหีบห่อ เราลองส่งอาร์กิวเมนต์ชนิด `int` ไปยัง `methodD` ซึ่งมีทั้งแบบที่พารามิเตอร์เป็น `long` และแบบที่เป็น `Integer` แต่ไม่มีแบบที่เป็น `int` ตรง ๆ ในกรณีนี้เมทอดที่เป็นชนิดที่กว้างขึ้นจะถูกเรียก แสดงว่า Java ให้ความสำคัญกับการแปลงชนิดให้กว้างขึ้นมาก่อนการแปลงเป็นคลาสหีบห่อ

> **หมายเหตุ** ในโลกของศาสตร์ด้านภาษาโปรแกรม การโอเวอร์โหลดจะเรียกอีกอย่างว่าการมีหลายรูปแบบแบบเฉพาะกิจ (ad hoc polymorphism) เนื่องจากทำให้เมทอดชื่อเดียวมีได้หลายรูปแบบ แต่จำนวนรูปแบบจำกัดเฉพาะตามรูปแบบที่เรากำหนดในพารามิเตอร์เท่านั้น ทั้งนี้ การเรียกการโอเวอร์โหลดว่าการมีหลายรูปแบบแบบเฉพาะกิจ จะจำกัดเฉพาะในวงสนทนาของศาสตร์ด้านภาษาโปรแกรม แต่ในบริบทของภาษาโปรแกรมเชิงวัตถุ เวลาที่เราใช้คำว่าการมีหลายรูปแบบ (polymorphism) ในที่นี้เราจะไม่ได้หมายถึงการโอเวอร์โหลด เราจะไม่เรียกการโอเวอร์โหลดว่าการมีหลายรูปแบบเพราะอาจจะทำให้สับสนกับแนวคิดการมีหลายรูปแบบอีกอย่างหนึ่งในการโปรแกรมเชิงวัตถุได้

## Nested classes

## การห่อหุ้ม

encapsulation

## UML

## References and aliasing

## null & NullPointerException

ตัวอย่างโค้ดหลาย ๆ ตัวอย่างที่แสดงในที่นี้อาจจะไม่มีการตรวจสอบข้อผิดพลาดมากนัก ทั้งนี้เนื่องจากต้องการเน้นประเด็นที่กำลังนำเสนอโดยไม่มีโค้ดอื่นที่ไม่ได้แสดงประเด็นเหล่านั้นปะปนอยู่มากเกินไป ในทางปฏิบัติแล้ว โค้ดควรจะต้องมีการตรวจสอบและจัดการข้อผิดพลาดอยู่เสมอ

## IllegalArgumentException

## Chaining method calls, fluent interface

## Static & dynamic initializers

In [7]:
import javax.swing.JOptionPane;

JOptionPane.showMessageDialog(null, "Test", "InfoBox", JOptionPane.INFORMATION_MESSAGE);

In [5]:
import java.util.Scanner;

Scanner input = new Scanner(System.in);
String s = input.next();
System.out.println(s);

EvalException: 

In [ ]:
public class Pair {
    protected int first;
    protected int second;
    
    public Pair(int first, int second) {
        this.first = first;
        this.second = second;
    }
    
    public int getFirst() {
        return first;
    }
    
    public int getSecond() {
        return second;
    }
    
    public void print() {
        System.out.println("(" + first + ", " + second + ")");
    }
}

public class NullablePair extends Pair {
    public void reset() {
        first = 0;
        second = 0;
    }
}